tarantool/ext_types/decimal.py

```diff
@@ -0,0 +1,226 @@
+
+    starts_with_zero = str_repr[first_digit_ind] == '0'
+
+    if (    (digit_count > TARANTOOL_DECIMAL_MAX_DIGITS + 1) or \
+            (digit_count == TARANTOOL_DECIMAL_MAX_DIGITS + 1 \
+            and not starts_with_zero)):
+        warn('Decimal encoded with loss of precision: ' + \
+             'Tarantool decimal supports a maximum of 38 digits.',
+             MsgpackWarning)
+        return False
+
+    return True
+
+def strip_decimal_str(str_repr, scale, first_digit_ind):
+    assert scale > 0
+    # Strip extra bytes
+    str_repr = str_repr[:TARANTOOL_DECIMAL_MAX_DIGITS + 1 + first_digit_ind]
+
+    str_repr.rstrip('0')
+    str_repr.rstrip('.')
+    # Do not strips zeroes before the decimal point.
+    return str_repr
+
+def encode(obj):
+    # Non-scientific string with trailing zeroes removed
+    str_repr = format(obj, 'f')
+
+    bytes_reverted = bytearray()
+
+    scale = 0
+    for i in range(len(str_repr)):
+        str_digit = str_repr[i]
+        if str_digit == '.':
+            scale = len(str_repr) - i - 1
+            break
+
+    if str_repr[0] == '-':
+        sign = '-'
+        first_digit_ind = 1
+    else:
+        sign = '+'
+        first_digit_ind = 0
+
+    if not check_valid_tarantool_decimal(str_repr, scale, first_digit_ind):
+        str_repr = strip_decimal_str(str_repr, scale, first_digit_ind)
+
+    bytes_reverted.append(get_mp_sign(sign))
+
+    digit_count = 0
+
+    for i in range(len(str_repr) - 1, first_digit_ind - 1, -1):
+        str_digit = str_repr[i]
+        if str_digit == '.':
+            scale = len(str_repr) - i - 1
+            continue
+
+        add_mp_digit(int(str_digit), bytes_reverted, digit_count)
+        digit_count = digit_count + 1
+
+    # Remove leading zeroes since they already covered by scale
+    for i in range(len(bytes_reverted)):
+        if bytes_reverted[i] != 0:
+            break
+        bytes_reverted.pop()
+
+    bytes_reverted.append(scale)
+
+    return bytes(bytes_reverted[::-1])
+
+
+def get_str_sign(nibble):
+    if nibble == 0x0a or nibble == 0x0c or nibble == 0x0e or nibble == 0x0f:
+        return '+'
+
+    if nibble == 0x0b or nibble == 0x0d:
+        return '-'
+
+    raise MsgpackError('Unexpected MP_DECIMAL sign nibble')
+
+def add_str_digit(digit, digits_reverted, scale):
+    if not (0 <= digit <= 9):
+        raise MsgpackError('Unexpected MP_DECIMAL digit nibble')
+
+    if len(digits_reverted) == scale:
+        digits_reverted.append('.')
+
+    digits_reverted.append(str(digit))
+
+def decode(data):
+    scale = data[0] 
+
+    sign = get_str_sign(data[-1] & 0x0f)
+
+    # Parse from tail since scale is counted from the tail.
+    digits_reverted = []
+
+    add_str_digit((data[-1] & 0xf0) >> 4, digits_reverted, scale)
+
+    for i in range(len(data) - 2, 0, -1):
+        add_str_digit(data[i] & 0x0f, digits_reverted, scale)
+        add_str_digit((data[i] & 0xf0) >> 4, digits_reverted, scale)
```

Comment:
It's a matter of taste, but it looks simply:
```suggestion
        add_str_digit((data[i]) >> 4, digits_reverted, scale)
```

Reply:
Thank you. It wasn't a matter of readability -- I just haven't worked with raw bytes for too long and missed out that it is excessive.

Reply:
Fixed